src/main/java/org/mule/extension/file/internal/source/DirectoryListener.java

```diff
@@ -331,4 +311,100 @@ private LocalDateTime getWatermarkTimestamp(LocalFileAttributes attributes) {
+                                                                                                                            final PollContext<InputStream, LocalFileAttributes> pollContext) {
+    try {
+      Map<String, Result<InputStream, LocalFileAttributes>> pendingFilesByTimeCheck = new HashMap<>();
+
+      if (filesToProcess.isEmpty()) {
+        return pendingFilesByTimeCheck;// no files to process
+      }
+
+      if (isNotNull(timeBetweenSizeCheckInMillis) && timeBetweenSizeCheckInMillis > 0) {
+        sleep(timeBetweenSizeCheckInMillis);
+      }
+
+      List<Result<InputStream, LocalFileAttributes>> currentList =
+          fileSystem.list(config, directoryPath.toString(), recursive, matcher, timeBetweenSizeCheckInMillis, null);
+      Map<String, Result<InputStream, LocalFileAttributes>> currentFilesMap = toMap(currentList);
+
+      if (currentList.isEmpty()) {
+        return pendingFilesByTimeCheck;// files already processed
+      }
+
+      Map<String, Result<InputStream, LocalFileAttributes>> filteredOldMap = filesToProcess.entrySet().stream()
+          .filter(entry -> currentFilesMap.containsKey(entry.getKey()))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+      processFiles(pollContext, pendingFilesByTimeCheck, currentFilesMap, filteredOldMap);
+      return pendingFilesByTimeCheck;
+    } catch (InterruptedException e) {
+      throw new MuleRuntimeException(createStaticMessage("Execution was interrupted while waiting to recheck file sizes"), e);
+    }
+  }
+
+  private void processFiles(PollContext<InputStream, LocalFileAttributes> pollContext,
+                            Map<String, Result<InputStream, LocalFileAttributes>> pendingFilesByTimeCheck,
+                            Map<String, Result<InputStream, LocalFileAttributes>> currentFilesMap,
+                            Map<String, Result<InputStream, LocalFileAttributes>> filteredOldMap) {
+    boolean sourceIsStopping = false;
+    for (final Map.Entry<String, Result<InputStream, LocalFileAttributes>> file : filteredOldMap.entrySet()) {
+      if (sourceIsStopping) {
+        closeResultQuietly(file.getValue());
+        continue;
+      }
+
+      Result<InputStream, LocalFileAttributes> oldInputStreamLocalFileAttributesResult = filteredOldMap.get(file.getKey());
+      Result<InputStream, LocalFileAttributes> currentInputStreamLocalFileAttributesResult = currentFilesMap.get(file.getKey());
+
+      LocalFileAttributes currentAttributes = currentInputStreamLocalFileAttributesResult.getAttributes().get();
+      LocalFileAttributes oldAttributes = oldInputStreamLocalFileAttributesResult.getAttributes().get();
+      if (matcher.test(currentAttributes)) {
+        if (currentAttributes.getSize() == oldAttributes.getSize()) {
+          PollContext.PollItemStatus status =
+              processFile(currentInputStreamLocalFileAttributesResult, currentAttributes, pollContext);
```

Comment:
I prefer to send `file.getValue()` instead of `currentInputStreamLocalFileAttributesResult`. It's easier to understand and it is the same, right?